emerging_optimizers/orthogonalized_optimizers/adaptive_muon.py

```diff
@@ -0,0 +1,242 @@
+            moment2.lerp_(orth_grad.square(), 1 - beta2)
+
+            # AdamW-style division: grad / (sqrt(moment2) + eps)
+            denom = moment2.sqrt() + eps
+            return orth_grad / denom
+
+        elif self.moment2_method == "normuon":
+            # NorMuon: Row or column-wise second moment
+            # Compute mean of squared gradients along one dimension based on shape
+            # Average along the longer dimension to preserve structure along shorter dim
+            avg_dim = -1 if orth_grad.shape[-2] >= orth_grad.shape[-1] else -2
+            v_mean = orth_grad.square().mean(dim=avg_dim, keepdim=True)
+
+            # Update second moment with EMA
+            moment2.lerp_(v_mean, 1 - beta2)
+
+            # NorMuon uses reciprocal square root with clamping
+            step_size = moment2.clamp_min(eps).rsqrt_()
+            return orth_grad * step_size
+
+        else:
+            raise TypeError(f"Invalid second moment method: {self.moment2_method}")
+
+    @torch.no_grad()  # type: ignore[misc]
+    @override
+    def step(self, closure: Callable[[], float] | None = None) -> float | None:
+        """Performs a single optimization step with second moment normalization.
+
+        Args:
+            closure: A closure that reevaluates the model and returns the loss.
+        """
+        if closure is None:
+            loss = None
+        else:
+            loss = closure()
+
+        for group in self.param_groups:
+            for p in group["params"]:
+                if p.dim() == 1:
+                    raise ValueError(f"{self.__class__.__name__} does not support 1D parameters")
+                grad = p.grad
+                if grad is None:
+                    continue
+                state = self.state[p]
+
+                # initialize momentum buffer and second moment buffer
+                if "momentum_buffer" not in state:
+                    state["momentum_buffer"] = torch.zeros_like(grad)
+                    self._initialize_moment2(state, grad)
+
+                exp_avg = state["momentum_buffer"]
+
+                self._apply_weight_decay_inplace(
+                    p,
+                    grad,
+                    group["lr"],
+                    group["weight_decay"],
+                )
+
+                # update momentum buffer with EMA of gradient
+                exp_avg.lerp_(grad, 1 - group["momentum_beta"])
+
+                # include nesterov momentum
+                if self.use_nesterov:
+                    grad = grad.lerp(exp_avg, group["momentum_beta"])
+                else:
+                    grad = exp_avg
+
+                with utils.fp32_matmul_precision(self.fp32_matmul_prec):
+                    grad = self.scaled_orthogonalize_fn(grad)
+
+                # Apply second moment normalization
+                grad = self._apply_moment2_normalization(
+                    orth_grad=grad,
```

Comment:
It is better to change output of line 228 to orth_grad, otherwise it is confusing here.

We've been sloppy in couple of other places. If it is trivial, can update altogether in this PR. or address in following PRs.